File: io-rb/mynow.rb
# router for my.nownownow.com
require_relative '../email.rb'
require_relative 'web.rb'
require 'net/http'
require 'uri'
require 'json'

WEBPDIR = '/var/www/html/nownownow.com/m/'
CDNHOST = DB.exec("select o.config('cdn-nnn-host')")[0]['config']
CDNUSER = DB.exec("select o.config('cdn-nnn-user')")[0]['config']
CDNPASS = DB.exec("select o.config('cdn-nnn-pass')")[0]['config']
CDNAPIK = DB.exec("select o.config('cdn-api-key')")[0]['config']

class MyNow
  def call(env)
    q = Rack::Request.new(env)
    kk = q.cookies['ok']

# GET /f
# ask email address or show message (?m=code) about form
    if q.get? && q.path_info == '/f'
      web2('mynow.authform', kk, q.params['m'])

# POST /f
# receive email address from GET /f form
# email templink then redirect to /f?m=_ message
    elsif q.post? && q.path_info == '/f'
      r = DB.exec("select head, body from mynow.authpost($1, $2)",
        [ kk, q.params['email'] ])[0]
      if r['head'].nil?  # means no errors
        Thread.new { sendemails }
      end
      web(r)

# GET /e
# clicked link from emailed templink always /e?t={tempcode}
# if cookie already set, redirect to /
# if tempcode found in DB, get id and person name
# show page either "not found" or form post login, posting tempcode and id#
    elsif q.get? && q.path_info == '/e'
      web2('mynow.welcome', kk, q.params['t'])

# POST /e
# temp_use: if tempcode and id match, use id to set cookie and delete tempcode
# set cookie and redirect to / (if it was wrong, / will send back to auth)
    elsif q.post? && q.path_info == '/e'
      web2('mynow.login', q.params['t'], q.params['i'])

# GET /z
# delete cookie and show "logged out" message
    elsif q.get? && q.path_info == '/z'
      web2('mynow.logout', kk)

# GET /
# home page asks their location
    elsif q.get? && q.path_info == '/'
      web2('mynow.whereru', kk)

# POST /where
# update their city/state/country
# redirect to /urls if successful or GET / if not
    elsif q.post? && q.path_info == '/where'
      web2('mynow.whereset', kk, q.params['city'], q.params['state'], q.params['country'])

# GET /urls
# form to add their URLs or, when done, link to GET /photo
    elsif q.get? && q.path_info == '/urls'
      web2('mynow.urls', kk)

# POST /urls
# add their URL and redirect to /urls for more
    elsif q.post? && q.path_info == '/urls'
      web2('mynow.urladd', kk, q.params['url'])

# POST /urls/main/[0-9]+
# set this URL as their main one and redirect to /urls
    elsif q.post? && (m = %r{\A/url/([1-9][0-9]*)/main\z}.match(q.path_info))
      web2('mynow.urlmain', kk, m[1].to_i)

# POST /urls/delete/[0-9]+
# delete this URL and redirect to /urls
    elsif q.post? && (m = %r{\A/url/([1-9][0-9]*)/delete\z}.match(q.path_info))
      web2('mynow.urldel', kk, m[1].to_i)

# GET /photo
# form to upload photo, and show uploaded 
# when done, link to GET /profile
    elsif q.get? && q.path_info == '/photo'
      web2('mynow.photo', kk)

# POST /photo
# receive uploaded photo: update now_profiles set photo=true
# if they uploaded photo, new name is public_id.webp
# vips copy #{tempfile} #{webp}[Q=100,strip]
# save to local path for display
# upload to CDN in background
# redirect to /photo
    elsif q.post? && q.path_info == '/photo'
      photo = q.params["photo"] || q.params[:photo]
      web({'head' => "303\r\nLocation: /photo"}) unless photo && photo[:tempfile]
      tempfile = photo[:tempfile].path
      web({'head' => "303\r\nLocation: /photo"}) unless File.exist?(tempfile)
      r = DB.exec("select code from mynow.photoset($1)", [kk])[0]
      webp = WEBPDIR + r['code'] + '.webp'
      system("vips copy #{tempfile} #{webp}[Q=100,strip]")
      if File.exist?(webp)
        Thread.new(webp) do |img| # make variable local to thread to be safe
          begin
            imgfile = img.gsub(WEBPDIR, '')
            url = URI.parse('https://ny.storage.bunnycdn.com/now3/' + imgfile)
            req = Net::HTTP::Put.new(url)
            req['AccessKey'] = CDNPASS
            req['Content-Type'] = 'image/webp'
            req.body = File.binread(img)
            Net::HTTP.start(url.host, url.port, use_ssl: true) do |http|
              http.request(req)
            end
            url = URI.parse('https://api.bunny.net/purge?url=https%3A%2F%2Fm.nownownow.com%2F' + imgfile)
            req = Net::HTTP::Post.new(url)
            req['AccessKey'] = CDNAPIK
            Net::HTTP.start(url.host, url.port, use_ssl: true) do |http|
              http.request(req)
            end
          rescue
            # oh well
          end
        end
      end
      web({'head' => "303\r\nLocation: /photo"})

# GET /profile
# form to ask the five profile questions
# when done, say thanks, that's the end
# ?edit1=title|liner|why|thought|red when complete to go back and edit one
# passing its null value - when not requested - is expected
    elsif q.get? && q.path_info == '/profile'
      web2('mynow.profile', kk, q.params['edit1'])

# POST /profile
# update the five profile questions
# redirect back to GET /profile
    elsif q.post? && q.path_info == '/profile'
      web2('mynow.profileset', kk, q.params['qcode'], q.params['answer'])

########### SITE CHECKER:

# POST /check/12/nodate || /check/12/gone
# done checking: update meta-info, send formletter, redirect to next check
    elsif q.post? && (m = %r{\A/check/([1-9][0-9]*)/(nodate|gone)\z}.match(q.path_info))
      web2('mynow.checkdone', kk, m[1], m[2])

# POST /check/123 with params[look4] and params[updated_at] in YYYY-MM-DD
# done checking: update meta-info, send formletter, redirect to next check
    elsif q.post? && (m = %r{\A/check/([1-9][0-9]*)\z}.match(q.path_info)) &&
      q.params['look4'] &&
      q.params['updated_at'] &&
      /2[0-9]{3}-[0-9]{2}-[0-9]{2}/ === q.params['updated_at']
      web2('mynow.checkupdate', kk, m[1], q.params['look4'], q.params['updated_at'])

# GET /check/123 - form to check that site
    elsif q.get? && (m = %r{\A/check/([1-9][0-9]*)\z}.match(q.path_info))
      web2('mynow.checkone', kk, m[1])

# GET /check - redirects to next /check/123
    elsif q.get? && q.path_info == '/check'
      web2('mynow.checknext', kk)

    else
      web({'head' => '404', 'body' => '?'})

    end
  end
end

File: mynow/mynow.go
package main

// so much of this code is just to handle the image uploading

import (
	"bytes"
	"image"
	_ "image/gif"
	_ "image/jpeg"
	_ "image/png"
	"log"
	"net/http"
	"os"

	"github.com/chai2010/webp"
	"sive.rs/sivers/internal/xx"
)

func main() {
	f, err := os.OpenFile("/tmp/mynow.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Printf("warning: couldn't open log file: %v", err)
	} else {
		log.SetOutput(f)
		defer f.Close()
	}

	if err := xx.InitDB(true); err != nil {
		log.Fatalf("InitDB %v", err)
	}
	defer xx.DB.Close()

	const WEBPDIR = "/var/www/html/nownownow.com/m/"
	var (
		CDNHOST string
		CDNPASS string
		CDNAPIK string
	)
	_ = xx.DB.QueryRow("select o.config('cdn-nnn-host')").Scan(&CDNHOST)
	_ = xx.DB.QueryRow("select o.config('cdn-nnn-pass')").Scan(&CDNPASS)
	_ = xx.DB.QueryRow("select o.config('cdn-api-key')").Scan(&CDNAPIK)

	mux := http.NewServeMux()

	mux.HandleFunc("GET /f", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		m := r.URL.Query().Get("m")
		xx.WebDB(w, r, "mynow.authform", kk, m)
	})

	mux.HandleFunc("POST /f", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		email := r.FormValue("email")
		xx.WebDB(w, r, "mynow.authpost", kk, email)
	})

	mux.HandleFunc("GET /e", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		t := r.URL.Query().Get("t")
		xx.WebDB(w, r, "mynow.welcome", kk, t)
	})

	mux.HandleFunc("POST /e", func(w http.ResponseWriter, r *http.Request) {
		t := r.FormValue("t")
		i := r.FormValue("i")
		xx.WebDB(w, r, "mynow.login", t, i)
	})

	mux.HandleFunc("GET /z", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		xx.WebDB(w, r, "mynow.logout", kk)
	})

	mux.HandleFunc("GET /{$}", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		xx.WebDB(w, r, "mynow.whereru", kk)
	})

	mux.HandleFunc("POST /where", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		city := r.FormValue("city")
		state := r.FormValue("state")
		country := r.FormValue("country")
		xx.WebDB(w, r, "mynow.whereset", kk, city, state, country)
	})

	mux.HandleFunc("GET /urls", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		xx.WebDB(w, r, "mynow.urls", kk)
	})

	mux.HandleFunc("POST /urls", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		url := r.FormValue("url")
		xx.WebDB(w, r, "mynow.urladd", kk, url)
	})

	mux.HandleFunc("POST /url/{id}/main", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		id := r.PathValue("id")
		xx.WebDB(w, r, "mynow.urlmain", kk, id)
	})

	mux.HandleFunc("POST /url/{id}/delete", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		id := r.PathValue("id")
		xx.WebDB(w, r, "mynow.urldel", kk, id)
	})

	mux.HandleFunc("GET /photo", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		xx.WebDB(w, r, "mynow.photo", kk)
	})

	mux.HandleFunc("GET /profile", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		edit1 := r.URL.Query().Get("edit1")
		xx.WebDB(w, r, "mynow.profile", kk, edit1)
	})

	mux.HandleFunc("POST /profile", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		qcode := r.FormValue("qcode")
		answer := r.FormValue("answer")
		xx.WebDB(w, r, "mynow.profileset", kk, qcode, answer)
	})

	mux.HandleFunc("POST /check/{id}/{action}", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		id := r.PathValue("id")
		action := r.PathValue("action")
		xx.WebDB(w, r, "mynow.checkdone", kk, id, action)
	})

	mux.HandleFunc("POST /check/{id}", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		id := r.PathValue("id")
		look4 := r.FormValue("look4")
		updatedAt := r.FormValue("updated_at")
		xx.WebDB(w, r, "mynow.checkupdate", kk, id, look4, updatedAt)
	})

	mux.HandleFunc("GET /check/{id}", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		id := r.PathValue("id")
		xx.WebDB(w, r, "mynow.checkone", kk, id)
	})

	mux.HandleFunc("GET /check", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)
		xx.WebDB(w, r, "mynow.checknext", kk)
	})

	// PHOTO UPLOAD
	mux.HandleFunc("POST /photo", func(w http.ResponseWriter, r *http.Request) {
		kk := xx.GetCookie(r)

		// get uploaded photo or redirect to /photo
		file, _, err := r.FormFile("photo")
		if err != nil {
			http.Redirect(w, r, "/photo", 303)
			return
		}
		defer file.Close()
		img, _, err := image.Decode(file)
		if err != nil {
			http.Redirect(w, r, "/photo", 303)
			return
		}

		// get code for naming webp file
		var code string
		err = xx.DB.QueryRow("select code from mynow.photoset($1)", kk).Scan(&code)
		if err != nil {
			xx.Oops(w, err)
			return
		}
		filename := code + ".webp"
		filepath := WEBPDIR + filename

		// convert their uploaded image to webp format
		var buf bytes.Buffer
		err = webp.Encode(&buf, img, &webp.Options{
			Lossless: false,
			Quality:  90,
		})
		if err != nil {
			xx.Oops(w, err)
			return
		}

		// new webp image is in buf.Bytes(). write to disk
		err = os.WriteFile(filepath, buf.Bytes(), 0644)
		if err != nil {
			xx.Oops(w, err)
			return
		}

		// in a goroutine, upload to CDN and purge old
		go func() {
			defer func() {
				if r := recover(); r != nil {
					log.Printf("panic in CDN upload: %v", r)
				}
			}()

			uploadURL := "https://" + CDNHOST + "/now3/" + filename
			req, err := http.NewRequest("PUT", uploadURL, bytes.NewReader(buf.Bytes()))
			if err != nil {
				log.Printf("ERROR creating PUT: %v\n", err)
				return
			}
			req.Header.Set("AccessKey", CDNPASS)
			req.Header.Set("Content-Type", "image/webp")
			client := &http.Client{Timeout: 30000000000} // 30 seconds in nanoseconds
			resp, err := client.Do(req)
			if err != nil {
				log.Printf("ERROR sending PUT: %v\n", err)
				return
			}
			resp.Body.Close()
			log.Printf("PUT %s Status: %d\n", uploadURL, resp.StatusCode)
			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
				return
			}

			purgeURL := "https://api.bunny.net/purge?url=https%3A%2F%2Fm.nownownow.com%2F" + filename
			req, err = http.NewRequest("POST", purgeURL, nil)
			if err != nil {
				log.Printf("ERROR creating POST: %v\n", err)
				return
			}
			req.Header.Set("AccessKey", CDNAPIK)
			resp, err = client.Do(req)
			if err != nil {
				log.Printf("ERROR sending POST: %v\n", err)
				return
			}
			resp.Body.Close()
			log.Printf("POST/PURGE %s Status: %d\n", purgeURL, resp.StatusCode)
			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
				log.Printf("CDN purge failed for %s", filename)
			}
		}()

	})

	log.Println("MyNow @ :2206")
	log.Fatal(http.ListenAndServe(":2206", xx.AuthExcept(mux, "/f", "/e", "/z")))
}
